### DISCUSSION TITLE
Add rendering for custom validation in CP publish forms

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#8934] [Ideas]
When adding custom validation rules, it would be helpful to have a place at the top of the element publish screen.

example: on `Entry::EVENT_BEFORE_SAVE`, I'm enforcing a workflow where certain transitions to certain entry types aren't allowed. However, adding errors `typeId` doesn't render anything in the template.

```
                    if (
                        $entry->type->handle === 'emailOrders' &&
                        $this->existingEntry->type->handle === 'orders'
                    ) {
                        // TODO: this error doesn't current show anywhere…
                        $entry->addError('typeId', 'Orders cannot be converted to email orders.');
                        $event->isValid = false;
                    }
```

I'm not necessarily advocating for `typeId` to render errors, but it would be nice if we could do something like `$entry->addError('custom', '…')`, and have it output errors at the top of the form.

**Comment by timkelty** [comment]
FWIW, as a stop-gap, I also tried `Craft::$app->getSession()->setError($message)`, but it looks at the point I was setting it, it looks like it gets overwritten by the generic "Couldn't save entry".

**Comment by brandonkelly** [comment]
Open to this, but in the meantime you should be able to do it using a custom Template element in your field layout, pointed to a template with this code:

```twig
{% if element and element.hasErrors('typeId') %}
  <div class="readable">
    <blockquote class="note warning">
      {{ element.getFirstError('typeId') }}
    </blockquote>
  </div>
{% endif %}
```

**Comment by timkelty** [comment]
ah, almost forgot about those! Thanks.

**Comment by brandonkelly** [comment]
Just went ahead and added error output for the `typeId` attribute as well.

**Comment by brandonkelly** [comment]
Craft 3.5.13 is out now with that change ✨

**Comment by timkelty** [comment]
UDABEST!
